lib/code-caller-docker.js

```diff
@@ -0,0 +1,703 @@
+    _checkState(allowedStates) {
+        if (allowedStates && !allowedStates.includes(this.state)) {
+            const allowedStatesList = allowedStates.map(String).join(', ');
+            return this._logError(`Expected DockerCaller to be in states [${allowedStatesList}] but actually have state ${String(this.state)}`);
+        }
+
+        let containerNull, callbackNull, timeoutIDNull;
+        if (this.state == CREATED) {
+            containerNull = true;
+            callbackNull = true;
+            timeoutIDNull = true;
+        } else if (this.state == WAITING) {
+            containerNull = false;
+            callbackNull = true;
+            timeoutIDNull = true;
+        } else if (this.state == IN_CALL) {
+            containerNull = false;
+            callbackNull = false;
+            timeoutIDNull = false;
+        } else if (this.state == EXITING) {
+            containerNull = false;
+            callbackNull = true;
+            timeoutIDNull = true;
+        } else if (this.state == EXITED) {
+            containerNull = true;
+            callbackNull = true;
+            timeoutIDNull = true;
+        } else {
+            return this._logError(`Invalid DockerCaller state: ${String(this.state)}`);
+        }
+
+        if (containerNull != null) {
+            if (containerNull && this.container != null) return this._logError(`DockerCaller state ${String(this.state)}: container should be null`);
+            if (!containerNull && this.container == null) return this._logError(`DockerCaller state ${String(this.state)}: container should not be null`);
+        }
+        if (callbackNull != null) {
+            if (callbackNull && this.callback != null) return this._logError(`DockerCaller state ${String(this.state)}: callback should be null`);
+            if (!callbackNull && this.callback == null) return this._logError(`DockerCaller state ${String(this.state)}: callback should not be null`);
+        }
+        if (timeoutIDNull != null) {
+            if (timeoutIDNull && this.timeoutID != null) return this._logError(`DockerCaller state ${String(this.state)}: timeoutID should be null`);
+            if (!timeoutIDNull && this.timeoutID == null) return this._logError(`DockerCaller state ${String(this.state)}: timeoutID should not be null`);
+        }
+
+        return true;
+    }
+}
+
+module.exports.DockerCaller = DockerCaller;
+
+/**
+ * If PrairieLearn dies unexpectedly, we may leave around temporary directories
+ * that should have been removed. This function will perform a best-effort
+ * attempt to clean them up, but will allow execution to continue if something
+ * fails. It'll check for any directories in the OS tmp directory that match the
+ * pattern of our tmp directory names, try to unmount them, and then remove the
+ * directories.
+ * 
+ * This function should be run on startup before workers are initialized.
+ */
+module.exports.cleanupMountDirectories = async function cleanupHostDirectories() {
```

Comment:
Rather than calling this directly from the init sequence in `server.js`, could we add an `init()` function in `code-caller-docker.js` that calls `cleanupMountDirectories()`, and call this `init()` from `server.js`?

We have a rough convention that any libraries needing initialization provide an `init()` function, so bringing up a server requires calling all of them, such as `aws.init()`, `cron.init()`, `workers.init()`, etc: https://github.com/PrairieLearn/PrairieLearn/blob/e9d7d530f6767ffb776e3019609d7edb85f2b32f/tests/helperServer.js#L42

For shutting down we are calling the corresponding function `close()`, so shutting down a server involves calling `freeformServer.close()`, `load.close()`, etc: https://github.com/PrairieLearn/PrairieLearn/blob/e9d7d530f6767ffb776e3019609d7edb85f2b32f/tests/helperServer.js#L144

I'll be the first to admit that we are often violating these naming conventions (hence `workers.finish()` and `cron.stop()`) but there is at least an aspiration to make this clean 😄

Reply:
For sure! More than happy to stick to established conventions 🙏

Reply:
Resolved in 695dfd3